pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/colocated/GreedyShuffleRewriteVisitor.java

```diff
@@ -190,23 +191,24 @@ public Set<ColocationKey> visitMailboxReceive(MailboxReceiveNode node, GreedyShu
       // distribution.
       ((MailboxSendNode) node.getSender()).setDistributionType(RelDistribution.Type.SINGLETON);
       return oldColocationKeys;
-    } else if (selector == null) {
+    } else if (distributionKeys == null) {
       return new HashSet<>();
     }
     // This means we can't skip shuffle and there's a partitioning enforced by receiver.
     int numPartitions = new HashSet<>(
         _dispatchablePlanMetadataMap.get(node.getPlanFragmentId()).getWorkerIdToServerInstanceMap().values()).size();
-    List<ColocationKey> colocationKeys = ((FieldSelectionKeySelector) selector).getColumnIndices().stream()
-        .map(x -> new ColocationKey(x, numPartitions, selector.hashAlgorithm())).collect(Collectors.toList());
+    List<ColocationKey> colocationKeys =
+        distributionKeys.stream().map(x -> new ColocationKey(x, numPartitions, KeySelector.DEFAULT_HASH_ALGORITHM))
```

Comment:
shall we make this algorithm configurable?

Reply:
This optimizer is triggered by a hint, and we are not using it. I believe the long term plan is to remove it. cc @walterddr

Reply:
i think for now we can assume that. the problem is we dont know how users will actually be using. having this configurable can be done in the future when needed.